### PR TITLE
Improve slice coverage in sweeps by adding N-dimensional slice support

### PR DESCRIPTION
**Pull request by sjameelTT**
### Ticket
#14100 (n-dimensional slice)
#13830 strided slice support for tiled
#13592 add slice to documentation

### Problem description
We need N-dimensional slice and slice on tiled inputs for generality. As well, we need to fix the failing test cases that are not related to tensor.

### What's changed
- Reworked strided slice kernel to make it N-dimensional
- Reworked slice invoke to make it generically N-dimensional
- Add slice to api.rst
- Fix DRAM read issue that caused bad pcc
- Add unit tests that track the remaining failing cases (with a skip) and the newly support inputs
- Add common code for TMs to encourage code reuse
- Remove bfloat8_b from sweeps as we're working on bfloat16 generality (since there are two times as many tiled tests, and tiled coverage was better, our numbers were looking better)
- Improved sweeps coverage to 75% from under 71% (71% included bfloat8_b). Tiled is now at 89.8% coverage
- Remaining failures are fixed with the tensor refactor as they all involve variants of either an odd inner dim input or output tensor for RM

### TODO
- Refactor SliceDeviceOperation to take ttnn::SimpleShape - I attempted to do this and got a lot of compile time reflection errors as it seems that SimpleShape lacks some attribute definitions (it may be because it stores shape as a vector, as opposed to the array in LegacyShape). Gave up for now since it's not an immediate priority.
- Improve perf in strided slice by making it multicore, moving work to writer
- Combine rm slice implementations
- Either wait for tensor or write a native strided slice implementation for tiled

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

**Review by ntarafdar — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp [hunk 1]
```diff
@@ -16,27 +16,33 @@ namespace py = pybind11;
 
 void bind_slice(py::module& module) {
     auto doc =
-        R"doc(
-            slice(input_tensor: ttnn.Tensor, slice_start: List[int[tensor rank], slice_end: List[int[tensor rank],  value: Union[int, float], *, Optional[ttnn.MemoryConfig] = None) -> ttnn.Tensor
+        R"doc(slice(input_tensor: ttnn.Tensor, slice_start: List[int[tensor rank], slice_end: List[int[tensor rank], slice_end: List[int[tensor rank], memory_config: Optional[MemoryConfig] = std::nullopt, queue_id: int = 0) -> ttnn.Tensor
```
Comment (yan-zaretskiy) [inline ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp:19]: We don't include the signature at the top of the docstring anymore.

File: ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp [hunk 2]
```diff
@@ -16,27 +16,33 @@ namespace py = pybind11;
...
+            Args:
+                input_tensor: Input Tensor.
+                slice_start: Start indices of input tensor. Values along each dim must be < input_tensor_shape[i].
+                slice_end: End indices of input tensor. Values along each dim must be < input_tensor_shape[i].
+                slice_step: (Optional[List[int[tensor rank]]) Step size for each dim. Default is None, which works out be 1 for each dimension.
 
-            .. code-block:: python
+            Keyword Args:
+                memory_config Memory Config of the output tensor
+                queue_id (Optional[uint8]) command queue id
```
Comment (yan-zaretskiy) [inline ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp:31]: Please see how we mark optional arguments, it should be `(uint8, optional)` if I am not mistaken.

File: tests/ttnn/unit_tests/operations/test_slice.py [hunk 3]
```diff
@@ -573,3 +599,190 @@ def test_ttnn_slice_optimized_shapes(input_shape, input_start, input_ends, layou
...
+    ttnn_output = ttnn.to_torch(ttnn_output)
+    assert_with_pcc(torch_output, ttnn_output, 0.99)
+
+
+def test_slice_7d_strided(device):
+    torch_input = torch.randn(1, 1, 1, 1, 1, 1, 256)
+    ttnn_input = ttnn.from_torch(torch_input, device=device, dtype=ttnn.bfloat16, layout=ttnn.ROW_MAJOR_LAYOUT)
+
+    torch_output = torch_input[..., 0:1, 0:1, 0:1, 0:1, 0:1, 0:256:2]
+    ttnn_output = ttnn_input[..., 0:1, 0:1, 0:1, 0:1, 0:1, 0:256:2]
```
Comment (yan-zaretskiy) [inline tests/ttnn/unit_tests/operations/test_slice.py:696]: Does this delegate to `ttnn.slice`?

**Review by tt-rkim — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp (hunk 2 above)
Comment (sjameelTT) [reply ttnn/cpp/ttnn/operations/data_movement/slice/slice_pybind.hpp:31]: Done, copied the one from permute which is uint8, optional like you said

File: tests/ttnn/unit_tests/operations/test_slice.py (hunk 3 above)
Comment (sjameelTT) [reply tests/ttnn/unit_tests/operations/test_slice.py:696]: Yep, core.py has some preprocessing that turns the slices into ttnn.slice compatible arguments

**Review by yan-zaretskiy — APPROVED** [review]

**Review by ayerofieiev-tt — APPROVED** [review]

**Review by TT-BrianLiu — APPROVED** [review]

**Review by esmalTT — APPROVED** [review]

**Pull request merged**